### PR TITLE
feat(inventory): software with special chars are poorly reconciled

### PR DESCRIPTION
**Pull request by stonebuzz**
From ```inventory``` when GLPI compare  software from database and from inventory file.

With ```inventory``` when GLPI compares the software of the database and those of the file of inventory, 

the method allowing to create the key of comparison does not take into account special characters encoded in database (encoded with ```htmlentities```)

From JSON

```json
            {
                "arch": "neutral",
                "folder": "C:\\Program Files\\WindowsApps\\SoundcloudLtd.SoundCloudforWindowsBeta_2.0.0.0_neutral__2xc63xn306dnw",
                "from": "uwp",
                "install_date": "2023-01-13",
                "name": "SoundCloud - Music & Songs",
                "publisher": "SoundCloud Global Limited & Co. KG",
                "system_category": "store",
                "version": "2.0.0.0"
            }
```  

from DB (see ```&#38;``` for ```&```)

```php
Array
  (
      [softid] => 3065
      [name] => SoundCloud - Music &#38; Songs
      [versionid] => 3331
      [version] => 2.0.0.0
      [arch] => neutral
      [entities_id] => 0
      [softwarecategories_id] => store
      [is_recursive] => 0
      [operatingsystems_id] => 0
      [manufacturers_id] => SoundCloud Global Limited &#38; Co. KG
  )
```

Generated comparison key from JSON
```
12a69d811b2a9eb60fbd52dfbe8fdbdc8f7b19ee$$$$2.0.0.0$$$$neutral$$$$SoundCloud Global Limited & Co. KG$$$$0$$$$0$$$$0
```

Generated comparison key from DB
```
e8688186f7ad446305b996140b146d36156b77b2$$$$2.0.0.0$$$$neutral$$$$SoundCloud Global Limited &#38; Co. KG$$$$0$$$$0$$$$0 
```

This Pr fix this

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/glpi-project/glpi-inventory-plugin/issues/296

### CODE REVIEW

File: src/Inventory/Asset/Software.php [hunk 1]
```diff
@@ -498,11 +499,12 @@ protected function getVersionKey($val, $softwares_id): string
      */
     protected function getFullCompareKey(\stdClass $val, bool $with_version = true): string
     {
+        global $DB;
         return $this->getCompareKey([
-            sha1($val->name),
+            sha1($DB->escape(Html::entity_decode_deep($val->name))),
```
Comment (cedric-anne) [inline src/Inventory/Asset/Software.php:504]:
```suggestion
            sha1(Sanitizer::sanitize($val->name)),
```

File: src/Inventory/Asset/Software.php [hunk 2]
```diff
@@ -498,11 +499,12 @@ protected function getVersionKey($val, $softwares_id): string
...
             $with_version ? strtolower($val->version) : '',
             strtolower($val->arch ?? ''),
-            $val->manufacturers_id,
+            $DB->escape(Html::entity_decode_deep($val->manufacturers_id)),
```
Comment (cedric-anne) [inline src/Inventory/Asset/Software.php:507]:
```suggestion
            Sanitizer::sanitize($val->manufacturers_id),
```

File: src/Inventory/Asset/Software.php [hunk 3]
```diff
@@ -518,8 +520,9 @@ protected function getFullCompareKey(\stdClass $val, bool $with_version = true):
      */
     protected function getSimpleCompareKey(\stdClass $val): string
     {
+        global $DB;
         return $this->getCompareKey([
-            sha1($val->name),
+            sha1($DB->escape(Html::entity_decode_deep($val->name))),
```
Comment (cedric-anne) [inline src/Inventory/Asset/Software.php:525]:
```suggestion
            sha1(Sanitizer::sanitize($val->name)),
```

File: src/Inventory/Asset/Software.php [hunk 4]
```diff
@@ -518,8 +520,9 @@ protected function getFullCompareKey(\stdClass $val, bool $with_version = true):
      */
     protected function getSimpleCompareKey(\stdClass $val): string
     {
+        global $DB;
```
Comment (cedric-anne) [inline src/Inventory/Asset/Software.php:523]:
```suggestion
```

File: src/Inventory/Asset/Software.php [hunk 5]
```diff
@@ -498,11 +499,12 @@ protected function getVersionKey($val, $softwares_id): string
      */
     protected function getFullCompareKey(\stdClass $val, bool $with_version = true): string
     {
+        global $DB;
```
Comment (cedric-anne) [inline src/Inventory/Asset/Software.php:502]:
```suggestion
```

File: src/Inventory/Asset/Software.php [hunk 6]
```diff
@@ -41,6 +41,7 @@
 use Entity;
 use Glpi\Inventory\Conf;
 use Glpi\Toolbox\Sanitizer;
+use Html;
```
Comment (cedric-anne) [inline src/Inventory/Asset/Software.php:44]:
```suggestion
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Comment by cedric-anne** [comment]
@stonebuzz Could you add a test case for this ? W will need to validate that #12451 will not break this specific case.

**Review by cedric-anne — COMMENTED** [review]
Almost all properties may have special chars.
Maybe it would be more efficient to handle `\stdClass` in `Sanitizer::sanitize()` and then use `$val = Sanitizer::sanitize()` on first line of `get*Key` methods.

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**